File: tests/fixtures/xpath_injection/javascript/safe_xpath_query.js
// Safe: user-supplied substring routed through the project-local
// `escapeXpath` helper before concatenation.  The sanitizer clears the
// XPATH_INJECTION cap so the sink does not fire.
const xpath = require('xpath');
const { DOMParser } = require('xmldom');

function escapeXpath(raw) {
    return raw.replace(/'/g, '&apos;').replace(/"/g, '&quot;');
}

function lookup(req, res) {
    const doc = new DOMParser().parseFromString('<root/>');
    const user = req.query.user;
    const safe = escapeXpath(user);
    const expr = "//user[name='" + safe + "']";
    const nodes = xpath.select(expr, doc);
    res.json(nodes);
}

module.exports = lookup;

File: tests/fixtures/xpath_injection/c/safe_xpath_query.c
/* Safe: project-local sanitize_xpath (matches the developer-named
 * `sanitize_*` Sanitizer rule) clears caps on the user value before it
 * reaches xmlXPathEvalExpression. */
#include <libxml/xpath.h>
#include <stdlib.h>

extern char *sanitize_xpath(const char *raw);

xmlXPathObjectPtr do_lookup(xmlXPathContextPtr ctx) {
    char *user_expr = getenv("USER_EXPR");
    char *safe = sanitize_xpath(user_expr);
    return xmlXPathEvalExpression((xmlChar *)safe, ctx);
}
